Stratis.Bitcoin/BlockPulling/QualityScore.cs

```diff
@@ -103,29 +92,28 @@ public void AddSample(IBlockPullerBehavior peer, long blockDownloadTimeMs, int b
 
             lock (this.lockObject)
             {
-                // If we reached the maximum number of samples, we need to remove oldest sample.
-                if (this.samplesCount == this.samples.Length)
-                {
-                    PeerSample oldSample = this.samples[this.samplesIndex];
+                // Add new sample to the mix.
+                PeerSample newSample = new PeerSample();
+                newSample.timePerKb = timePerKb;
+                newSample.peer = peer;
+
+                if (this.peerReferenceCounter.ContainsKey(peer)) this.peerReferenceCounter[peer]++;
+                else this.peerReferenceCounter.Add(peer, 1);
+
+                PeerSample oldSample;
+                if (this.samples.Add(newSample, out oldSample))
```

Comment:
you can even do `if (this.samples.Add(newSample, out PeerSample oldSample))` now, how cool is that?